File: src/tabshifter/Actions.java
package tabshifter;

import com.intellij.openapi.actionSystem.*;
import com.intellij.openapi.fileEditor.ex.*;
import com.intellij.openapi.project.*;
import org.jetbrains.annotations.*;

import static tabshifter.Directions.*;

public class Actions {
	public static class ShiftLeft extends AnAction implements DumbAware {
		@Override public void actionPerformed(@NotNull AnActionEvent event) {
			tabShifter(event).moveTab(left);
		}
	}
	public static class ShiftUp extends AnAction implements DumbAware {
		@Override public void actionPerformed(@NotNull AnActionEvent event) {
			tabShifter(event).moveTab(up);
		}
	}
	public static class ShiftRight extends AnAction implements DumbAware {
		@Override public void actionPerformed(@NotNull AnActionEvent event) {
			tabShifter(event).moveTab(right);
		}
	}
	public static class ShiftDown extends AnAction implements DumbAware {
		@Override public void actionPerformed(@NotNull AnActionEvent event) {
			tabShifter(event).moveTab(down);
		}
	}

	public static class MoveFocusLeft extends AnAction implements DumbAware {
		@Override public void actionPerformed(@NotNull AnActionEvent event) {
			tabShifter(event).moveFocus(left);
		}
	}
	public static class MoveFocusUp extends AnAction implements DumbAware {
		@Override public void actionPerformed(@NotNull AnActionEvent event) {
			tabShifter(event).moveFocus(up);
		}
	}
	public static class MoveFocusRight extends AnAction implements DumbAware {
		@Override public void actionPerformed(@NotNull AnActionEvent event) {
			tabShifter(event).moveFocus(right);
		}
	}
	public static class MoveFocusDown extends AnAction implements DumbAware {
		@Override public void actionPerformed(@NotNull AnActionEvent event) {
			tabShifter(event).moveFocus(down);
		}
	}


	public static class StretchRight extends AnAction implements DumbAware {
		@Override public void actionPerformed(@NotNull AnActionEvent event) {
			tabShifter(event).stretchSplitter(right);
		}
	}
	public static class StretchLeft extends AnAction implements DumbAware {
		@Override public void actionPerformed(@NotNull AnActionEvent event) {
			tabShifter(event).stretchSplitter(left);
		}
	}
	public static class StretchUp extends AnAction implements DumbAware {
		@Override public void actionPerformed(@NotNull AnActionEvent event) {
			tabShifter(event).stretchSplitter(up);
		}
	}
	public static class StretchDown extends AnAction implements DumbAware {
		@Override public void actionPerformed(@NotNull AnActionEvent event) {
			tabShifter(event).stretchSplitter(down);
		}
	}

	public static class EqualSizeSplit extends AnAction implements DumbAware {
		@Override public void actionPerformed(@NotNull AnActionEvent event) {
			tabShifter(event).equalSizeSplitter();
		}
	}

	public static class ToggleMaximizeRestore extends AnAction implements DumbAware {
		@Override public void actionPerformed(@NotNull AnActionEvent event) {
			tabShifter(event).toggleMaximizeRestoreSplitter();
		}
	}

	private static TabShifter tabShifter(AnActionEvent event) {
		Project project = event.getProject();
		if (project == null) return TabShifter.none;

		FileEditorManagerEx editorManager = FileEditorManagerEx.getInstanceEx(project);
		if (editorManager == null || editorManager.getAllEditors().length == 0) return TabShifter.none;

		return new TabShifter(new Ide(editorManager, project));
	}
}

File: src/tabshifter/Actions.kt
package tabshifter

import com.intellij.openapi.actionSystem.AnAction
import com.intellij.openapi.actionSystem.AnActionEvent
import com.intellij.openapi.fileEditor.ex.FileEditorManagerEx
import com.intellij.openapi.project.DumbAware

object Actions {
    private fun tabShifter(event: AnActionEvent): TabShifter? {
        val project = event.project ?: return null
        val editorManager = FileEditorManagerEx.getInstanceEx(project) ?: return null
        if (editorManager.allEditors.isEmpty()) return null
        return TabShifter(Ide(editorManager, project))
    }

    class ShiftLeft: AnAction(), DumbAware {
        override fun actionPerformed(event: AnActionEvent) {
            tabShifter(event)?.moveTab(left)
        }
    }

    class ShiftUp: AnAction(), DumbAware {
        override fun actionPerformed(event: AnActionEvent) {
            tabShifter(event)?.moveTab(up)
        }
    }

    class ShiftRight: AnAction(), DumbAware {
        override fun actionPerformed(event: AnActionEvent) {
            tabShifter(event)?.moveTab(right)
        }
    }

    class ShiftDown: AnAction(), DumbAware {
        override fun actionPerformed(event: AnActionEvent) {
            tabShifter(event)?.moveTab(down)
        }
    }

    class MoveFocusLeft: AnAction(), DumbAware {
        override fun actionPerformed(event: AnActionEvent) {
            tabShifter(event)?.moveFocus(left)
        }
    }

    class MoveFocusUp: AnAction(), DumbAware {
        override fun actionPerformed(event: AnActionEvent) {
            tabShifter(event)?.moveFocus(up)
        }
    }

    class MoveFocusRight: AnAction(), DumbAware {
        override fun actionPerformed(event: AnActionEvent) {
            tabShifter(event)?.moveFocus(right)
        }
    }

    class MoveFocusDown: AnAction(), DumbAware {
        override fun actionPerformed(event: AnActionEvent) {
            tabShifter(event)?.moveFocus(down)
        }
    }

    class StretchRight: AnAction(), DumbAware {
        override fun actionPerformed(event: AnActionEvent) {
            tabShifter(event)?.stretchSplitter(right)
        }
    }

    class StretchLeft: AnAction(), DumbAware {
        override fun actionPerformed(event: AnActionEvent) {
            tabShifter(event)?.stretchSplitter(left)
        }
    }

    class StretchUp: AnAction(), DumbAware {
        override fun actionPerformed(event: AnActionEvent) {
            tabShifter(event)?.stretchSplitter(up)
        }
    }

    class StretchDown: AnAction(), DumbAware {
        override fun actionPerformed(event: AnActionEvent) {
            tabShifter(event)?.stretchSplitter(down)
        }
    }

    class EqualSizeSplit: AnAction(), DumbAware {
        override fun actionPerformed(event: AnActionEvent) {
            tabShifter(event)?.equalSizeSplitter()
        }
    }

    class ToggleMaximizeRestore: AnAction(), DumbAware {
        override fun actionPerformed(event: AnActionEvent) {
            tabShifter(event)?.toggleMaximizeRestoreSplitter()
        }
    }
}